python/pyiceberg/table/__init__.py

```diff
@@ -841,3 +868,239 @@ def to_ray(self) -> ray.data.dataset.Dataset:
         import ray
 
         return ray.data.from_arrow(self.to_arrow())
+
+
+class UpdateSchema:
+    def __init__(self, schema: Schema, table: Optional[Table] = None, last_column_id: Optional[int] = None):
+        self._table = table
+        self._schema = schema
+        if last_column_id:
+            self._last_column_id = itertools.count(last_column_id + 1)
+        else:
+            self._last_column_id = itertools.count(schema.highest_field_id + 1)
+
+        self._identifier_field_names = schema.column_names
+        self._adds: Dict[int, List[NestedField]] = {}
+        self._added_name_to_id: Dict[str, int] = {}
+        self._id_to_parent: Dict[int, str] = {}
+        self._allow_incompatible_changes: bool = False
+        self._case_sensitive: bool = True
+
+    def __exit__(self, _: Any, value: Any, traceback: Any) -> None:
```

Comment:
One question:

For:
```
with table.transaction() as transaction:
   transaction.update_schema().add_column("x1", IntegerType(), "doc").add_column("x2", IntegerType(), "doc").commit()
# now the commit happens because the transaction closes
```

There is no need to return the table, and there is no way to return the refreshed table.

Reply:
The reference to the table is refreshed:
```python
        table_update_response = self._table.catalog._commit_table(  # pylint: disable=W0212
            CommitTableRequest(
                identifier=self._table.identifier[1:],
                updates=[
                    AddSchemaUpdate(schema=new_schema, last_column_id=new_schema.highest_field_id),
                    SetCurrentSchemaUpdate(schema_id=-1),
                ],
            )
        )

        self._table.metadata = table_update_response.metadata
        self._table.metadata_location = table_update_response.metadata_location
```

So strictly it is not needed, and the table is already updated. I noticed this when checking the integration tests. With the context manager this is indeed not possible, I'm also fine with leaving this out.